packages/medusa/src/utils/authenticate-middleware.ts

```diff
@@ -0,0 +1,79 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { IAuthModuleService } from "@medusajs/types"
+import { NextFunction, RequestHandler } from "express"
+import { MedusaRequest, MedusaResponse } from "../types/routing"
+
+type MedusaSession = {
+  auth: {
+    [authScope: string]: {
+      user_id: string
+    }
+  }
+}
+
+type AuthType = "session" | "bearer"
+
+export default (
+  authScope: string,
+  authType: AuthType | AuthType[],
+  options: { allowUnauthenticated?: boolean } = {}
+): RequestHandler => {
+  return async (
+    req: MedusaRequest,
+    res: MedusaResponse,
+    next: NextFunction
+  ): Promise<void> => {
+    const authTypes = Array.isArray(authType) ? authType : [authType]
+    const authModule = req.scope.resolve<IAuthModuleService>(
+      ModuleRegistrationName.AUTH
+    )
+
+    // @ts-ignore
+    const session: MedusaSession = req.session || {}
+
+    if (authTypes.includes("session")) {
+      if (session.auth && session.auth[authScope]) {
+        const authUser = await authModule.retrieveAuthUser(
+          session.auth[authScope].user_id
+        )
```

Comment:
thought: if this fails it should probably result in a 401 and not a 404 so maybe worth handling that error